File: app/src/main/java/me/ycdev/android/devtools/sampler/SampleTaskInfo.java
package me.ycdev.android.devtools.sampler;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.FileWriter;
import java.util.ArrayList;

import me.ycdev.android.arch.utils.AppLogger;

class SampleTaskInfo {
    private static final String TAG = "SampleTaskInfo";

    private static final String KEY_PKG_COUNT = "pkg_count";
    private static final String KEY_PKG_NAME_PREFIX = "pkg_";
    private static final String KEY_SAMPLE_INTERVAL = "interval";
    private static final String KEY_SAMPLE_PERIOD = "period";
    private static final String KEY_START_TIME = "start_time";
    private static final String KEY_SAMPLE_CLOCK_TIME = "sample_clock_time";
    private static final String KEY_SAMPLE_COUNT = "sample_count";

    public ArrayList<String> pkgNames;
    public int sampleInterval; // in seconds
    public int samplePeriod; // in minutes
    public long startTime;
    public long sampleClockTime;
    public int sampleCount;

    public boolean isSampling;
    public ArrayList<FileWriter> fileWriters;
    public AppsSetStat preAppsSetStat;

    public String backupTaskInfo() {
        try {
            JSONObject json = new JSONObject();
            final int N = pkgNames.size();
            json.put(KEY_PKG_COUNT, N);
            for (int i = 0; i < N; i++) {
                json.put(KEY_PKG_NAME_PREFIX + i, pkgNames.get(i));
            }
            json.put(KEY_SAMPLE_INTERVAL, sampleInterval);
            json.put(KEY_SAMPLE_PERIOD, samplePeriod);
            json.put(KEY_START_TIME, startTime);
            json.put(KEY_SAMPLE_CLOCK_TIME, sampleClockTime);
            json.put(KEY_SAMPLE_COUNT, sampleCount);
            return json.toString();
        } catch (JSONException e) {
            AppLogger.w(TAG, "failed to create sample task info backup");
        }
        return null;
    }

    public static SampleTaskInfo restoreTaskInfo(String backup) {
        try {
            JSONObject json = new JSONObject(backup);
            SampleTaskInfo taskInfo = new SampleTaskInfo();
            taskInfo.pkgNames = new ArrayList<>();
            final int N = json.getInt(KEY_PKG_COUNT);
            for (int i = 0; i < N; i++) {
                taskInfo.pkgNames.add(json.getString(KEY_PKG_NAME_PREFIX + i));
            }
            taskInfo.sampleInterval = json.getInt(KEY_SAMPLE_INTERVAL);
            taskInfo.samplePeriod = json.getInt(KEY_SAMPLE_PERIOD);
            taskInfo.startTime = json.getLong(KEY_START_TIME);
            taskInfo.sampleClockTime = json.getLong(KEY_SAMPLE_CLOCK_TIME);
            taskInfo.sampleCount = json.getInt(KEY_SAMPLE_COUNT);
            return taskInfo;
        } catch (JSONException e) {
            AppLogger.w(TAG, "failed to restore sample task info from backup");
        }
        return null;
    }
}

File: app/src/main/java/me/ycdev/android/devtools/sampler/SampleTaskInfo.kt
package me.ycdev.android.devtools.sampler

import org.json.JSONException
import org.json.JSONObject
import timber.log.Timber
import java.io.FileWriter
import java.util.ArrayList

class SampleTaskInfo {
    val pkgNames: ArrayList<String> = arrayListOf()
    var sampleInterval = 0 // in seconds
    var samplePeriod = 0 // in minutes
    var startTime: Long = 0
    var sampleClockTime: Long = 0
    var sampleCount = 0
    var isSampling = false
    var fileWriters: ArrayList<FileWriter?>? = null
    var preAppsSetStat: AppsSetStat? = null

    fun backupTaskInfo(): String? {
        try {
            val json = JSONObject()
            json.put(KEY_PKG_COUNT, pkgNames.size)
            for (i in 0 until pkgNames.size) {
                json.put(KEY_PKG_NAME_PREFIX + i, pkgNames[i])
            }
            json.put(KEY_SAMPLE_INTERVAL, sampleInterval)
            json.put(KEY_SAMPLE_PERIOD, samplePeriod)
            json.put(KEY_START_TIME, startTime)
            json.put(KEY_SAMPLE_CLOCK_TIME, sampleClockTime)
            json.put(KEY_SAMPLE_COUNT, sampleCount)
            return json.toString()
        } catch (e: JSONException) {
            Timber.tag(TAG).w(e, "failed to create sample task info backup")
        }
        return null
    }

    companion object {
        private const val TAG = "SampleTaskInfo"

        private const val KEY_PKG_COUNT = "pkg_count"
        private const val KEY_PKG_NAME_PREFIX = "pkg_"
        private const val KEY_SAMPLE_INTERVAL = "interval"
        private const val KEY_SAMPLE_PERIOD = "period"
        private const val KEY_START_TIME = "start_time"
        private const val KEY_SAMPLE_CLOCK_TIME = "sample_clock_time"
        private const val KEY_SAMPLE_COUNT = "sample_count"

        fun restoreTaskInfo(backup: String): SampleTaskInfo? {
            try {
                val json = JSONObject(backup)
                val taskInfo = SampleTaskInfo()
                val count = json.getInt(KEY_PKG_COUNT)
                for (i in 0 until count) {
                    taskInfo.pkgNames.add(json.getString(KEY_PKG_NAME_PREFIX + i))
                }
                taskInfo.sampleInterval = json.getInt(KEY_SAMPLE_INTERVAL)
                taskInfo.samplePeriod = json.getInt(KEY_SAMPLE_PERIOD)
                taskInfo.startTime = json.getLong(KEY_START_TIME)
                taskInfo.sampleClockTime = json.getLong(KEY_SAMPLE_CLOCK_TIME)
                taskInfo.sampleCount = json.getInt(KEY_SAMPLE_COUNT)
                return taskInfo
            } catch (e: JSONException) {
                Timber.tag(TAG).w(e, "failed to restore sample task info from backup")
            }
            return null
        }
    }
}